src/compose/app.ts

```diff
@@ -647,38 +742,70 @@ export class App {
 		);
 	}
 
-	private generateContainerStep(current: Service, target: Service) {
-		// if the services release doesn't match, then rename the container...
+	private generateContainerStep(
+		current: Service,
+		target: Service,
+		appsToLock: AppsToLockMap,
+		targetServices: Service[],
+		servicesLocked: boolean,
+	): CompositionStep[] {
+		// Update container metadata if service release has changed
 		if (current.commit !== target.commit) {
-			return generateStep('updateMetadata', { current, target });
+			// Only update service once locked or if legacy container
+			const shouldSkipLock = checkTruthy(
+				current.config.labels['io.balena.legacy-container'],
```

Comment:
What are legacy containers? Can there still be legacy containers at this point?

Reply:
According to https://github.com/balena-os/balena-supervisor/blob/addffa3dd79ac27ba69b93606e9e4e3fa91c435b/src/compose/app.ts#L265 and https://github.com/balena-os/balena-supervisor/blob/6217546894e25c8216d305f231876fc888a6e407/src/migrations/20171129064057_multicontainer.js#L12, looks like it's a container from before we added multicontainer support. There's a possibility a Supervisor could be running a legacy app, perhaps best to keep this in

Reply:
Can legacy containers have locks? If they cannot why should we explicitely skip them? I don't want us to add code just in case

Reply:
We had update locks as far back as Supervisor v4.0.0 according to our docs, and Supervisor added multicontainer support in 7.0.0. So I think the answer to that is yes, legacy containers can have update locks. However, given how old this is and the difficulty of reproduction, if you feel strongly about removing this, that SGTM

Reply:
Is not that I feel strongly about it, I'm just trying to understand the logic, in order for a user to still have legacy containers they would have to still be running the same containers from when they updated to v7, meaning no release updates from them, no OS update, and they have not yet jumped to a supervisor that recreates their containers. 

Assuming that can happen, I still don't understand why we would skip lock for those specific containers

Reply:
We discussed this further in our Supervisor checkin, and agreed it would be safe to remove this `skipLock` check. The chance of a device on such an old OS / Supervisor running an old release is really small, and if such a device were to appear on support, we should recommend a Supervisor / OS upgrade for them first thing anyway